internal/api/token.go

```diff
@@ -140,7 +141,31 @@ func (a *API) ResourceOwnerPasswordGrant(ctx context.Context, w http.ResponseWri
 		return internalServerError("Database error querying schema").WithInternalError(err)
 	}
 
-	if user.IsBanned() || !user.Authenticate(ctx, params.Password) {
+	if user.IsBanned() {
+		return oauthError("invalid_grant", InvalidLoginMessage)
+
+	}
+	isValidPassword := user.Authenticate(ctx, params.Password)
+	if config.Hook.PasswordVerificationAttempt.Enabled {
+
+		input := hooks.PasswordVerificationAttemptInput{
+			UserID: user.ID,
+			Valid:  isValidPassword,
+		}
+		output := hooks.PasswordVerificationAttemptOutput{}
+		err := a.invokeHook(ctx, &input, &output)
+		if err != nil {
+			return err
+		}
+
+		if output.Decision == hooks.HookRejection {
+			if output.Message == "" {
```

Comment:
We don't logout the user as the user typically isn't already logged in when they use the password token grant

Reply:
Ah but you do log put the user elsewhere. Maybe we shpuld have an option in the putput to deal with this?